### DISCUSSION TITLE
persistSession:false causes NodeJS server to not return data on query

### DISCUSSION THREAD

**Discussion by Topener** [supabase/supabase#16571] [Questions]
I've been building a NodeJS API for some time, and recently updated supabase to the latest version. 

I used to have to have just this line:

```js
const supabase = createClient(process.env.SUPABASE_HOST, process.env.SUPABASE_KEY);
```

This worked fine, but started throwing an error recently, I'm assuming a stricter check/setting for when it's a NodeJS environment and there's no way to actually store the session
```
No storage option exists to persist the session, which may result in unexpected behavior when using auth.
        If you want to set persistSession to true, please provide a storage option or you may set persistSession to false to disable this warning.
```

However, when I turned it into this code:
```js
const supabase = createClient(process.env.SUPABASE_HOST, process.env.SUPABASE_KEY, {
    auth: {
        autoRefreshToken: false,
        persistSession: false,
        detectSessionInUrl: false,
    },
});
```

I kept getting 0 results on this query after I refreshed the token, even though the API is using the admin key and has access to everything.

```js
const { data, error } = await supabase.from("user_details").select("*").eq("user", userId).single();
```

The token refreshing
```js
const { data, error } = await supabase.auth.refreshSession({ refresh_token });
```

Setting `persistSession: true` fixed the issue, but now throws the error again. 

I'm wondering what I'm doing wrong? Should I be refreshing the token differently for the user doing the API call?

**Comment by GaryAustin1** [comment]
Not really sure all that is going on here, but...
With persistSession:true with no local storage, gotrue would clear out the session info on things like getSession and refreshSession.
In your case with persistSession:true, if that happens the user session gets cleared and the service_role (I assume that is what you mean by admin) key gets passed thru instead.  So you bypass RLS and just rely on your filter.

With persistSession:false your refresh session sets the session to a real session and your call goes out as an authenticated user
and your RLS must be met.  If the RLS is not met then you would get back empty array.

Maybe?  The warning is there because of the getSesssion() and I think refreshSession() clearing the session with no local storage. https://github.com/supabase/gotrue-js/issues/539


Note:  You really should only use service_role when you do not intend to have a user session to avoid this security hole on a mistake.  Anon key if passed thru would still obey your RLS.

**Comment by Topener** [comment]
Hi, thanks for your input. I do not store user sessions on the server-side, however, I do facilitate clients through an API to refresh their token, etc. We don't have Supabase on the client so everything runs through our own backend. 

So, we've got an endpoint to refresh the token, we've got a server-side check with supabase if the token is valid, etc. All RLS is validated server-side as well.

**Reply by GaryAustin1** [reply]
So with anon key and persistSession:false you should be able to refresh token.  If it does not work then my guess is you are not meeting RLS and in the way you had it you were bypassing RLS.

**Reply by Topener** [reply]
The refresh is working, but then I'm getting 0 results on this query,
```
const { data, error } = await supabase.from("user_details").select("*").eq("user", userId).single();
```
But after a server restart, I'm testing it locally but also reproducable on a server, the same query works fine.

The `userId` is correct, and is coming from the refreshed JWT token check, but somehow the query itself says 0 results, and I'm not sure where to look next, only that is was caused by setting it to false.

**Reply by GaryAustin1** [reply]
Not sure.  If you change to anon key instead of service_role key for createClient() and then have persistSession:true does it still work?  If not then it is service_role saving you right now and the user session is not working properly.

**Reply by Topener** [reply]
Ok you got me on the right path, turns out I did have RLS enabled on every table, and until now it never causes any issues. I did use the `service_role`, but the anon key worked the same. No clue why. But disabling RLS fixed it. I never use it in public anyways.
